File: WowUp.WPF/Utilities/TocParser.cs
using System.Text.RegularExpressions;
using WowUp.WPF.Models;

namespace WowUp.WPF.Utilities
{
    public class TocParser
    {
        private readonly string _tocText;
        private readonly Regex ColorRegex = new Regex(@"\|[a-zA-Z0-9]{9}");
        private readonly Regex NewLineRegex = new Regex(@"\|r");

        public string Interface => GetValue("Interface");
        public string Title => GetValue("Title");
        public string Author => GetValue("Author");
        public string Website => GetValue("X-Website");
        public string Version => GetValue("Version");
        public string PartOf => GetValue("X-Part-Of");
        public string Category => GetValue("X-Category");
        public string Localizations => GetValue("X-Localizations");
        public string CurseProjectId => GetValue("X-Curse-Project-ID");
        public string WowInterfaceId => GetValue("X-WoWI-ID");
        public string Dependencies => GetValue("Dependencies");

        public Toc Toc => new Toc
        {
            Author = Author,
            Category = Category,
            Interface = Interface,
            Localizations = Localizations,
            PartOf = PartOf,
            Title = Title,
            Version = Version,
            Website = Website,
            Dependencies = Dependencies,
            CurseProjectId = CurseProjectId,
            WowInterfaceId = WowInterfaceId
        };

        public TocParser(string tocText)
        {
            _tocText = tocText;
        }

        private string GetValue(string key)
        {
            var regex = new Regex($"^## {key}:(.*?)$", RegexOptions.Multiline);
            var match = regex.Match(_tocText);

            if (match != null && match.Groups.Count == 2)
            {
                var value = match?.Groups[1].Value.Trim();
                value = StripEncodedChars(value);

                return value;
            }

            return string.Empty;
        }

        private string StripEncodedChars(string value)
        {
            var str = StripColorChars(value);
            str = StripNewLineChars(str);

            return str;
        }

        private string StripColorChars(string value)
        {
            var str = value;
            Match match;
            while ((match = ColorRegex.Match(str)).Success)
            {
                str = str.Replace(match.Value, "");
            }

            return str;
        }

        private string StripNewLineChars(string value)
        {
            var str = value;
            Match match;
            while ((match = NewLineRegex.Match(str)).Success)
            {
                str = str.Replace(match.Value, "");
            }

            return str;
        }
    }
}

File: wowup-electron/src/app/utils/toc-parser.ts
import { promisify } from 'util';
import { FileUtils } from './file.utils';
import * as fs from 'fs';
import { Toc } from 'app/models/wowup/toc';

const readFileAsync = promisify(fs.readFile);

export class TocParser {

  private readonly _tocPath: string;
  private _tocText: string;

  private get newLineRegex() {
    return;
  }

  constructor(tocPath: string) {
    this._tocPath = tocPath;
  }

  public async parse(): Promise<Toc> {
    this._tocText = fs.readFileSync(this._tocPath, { encoding: 'utf-8' })

    return {
      author: this.getValue('Author'),
      curseProjectId: this.getValue('X-Curse-Project-ID'),
      interface: this.getValue('Interface'),
      title: this.getValue('Title'),
      website: this.getValue('Website'),
      version: this.getValue('Version'),
      partOf: this.getValue('X-Part-Of'),
      category: this.getValue('X-Category'),
      localizations: this.getValue('X-Localizations'),
      wowInterfaceId: this.getValue('X-WoWI-ID'),
      dependencies: this.getValue('Dependencies'),
    };
  }

  public parseMetaData(): string[] {
    this._tocText = fs.readFileSync(this._tocPath, { encoding: 'utf-8' })

    return this._tocText
      .split('\n')
      .filter(line => line.trim().startsWith('## '));
  }

  private getValue(key: string): string {
    const match = new RegExp(`^## ${key}:(.*?)$`, 'm').exec(this._tocText);

    if (!match || match.length !== 2) {
      return '';
    }

    return this.stripEncodedChars(match[1].trim());
  }

  private stripEncodedChars(value: string) {
    let str = this.stripColorChars(value);
    str = this.stripNewLineChars(str);

    return str;
  }

  private stripColorChars(value: string) {
    return value.replace(/\|[a-zA-Z0-9]{9}/g, '');
  }

  private stripNewLineChars(value: string) {
    return value.replace(/\|r/g, '');
  }
}